File: ml/tests/matlab/logistic_regression/test1_tumors.m
% Mihail Dunaev
% May 2017
%
% Simple test that tries to predict if a tumor is malignant (y = 1)
%   or benign (y = 0) based on the tumor size (one feature). It uses
%   linear logistic regression to do that. Data is in 'tumors.in' in
%   input dir.
%
% Results
%   threshold value for x (x0) is at -w(2)/w(1)
%   fminsearch w = 1.9564, -85.8940, cost = 0, x0 = 43.9041
%   fminunc w = 0.8260, -41.5848, cost = 1.0601e-06, x0 = 50.3448
%   fminunc with grad w = 0.6321, -27.3605, cost = 1.0773e-06, x0 = 43.2851
%   own gradient descent w = 1.8705, -70.4174, cost = 2.9976e-14,
%       x0 = 37.6463 (alpha = 7)

function test1_tumors()
    addpath('core/');
    [xs, ys] = read_data('../../input/tumors.in');
    
    syms t;
    w0 = [0; -5];
    fun = @(t)(cost_with_grad(t, xs, ys));
    w = train(w0, fun);
    
    % predict some values
    xpred = [7 43 44 70]';
    ypred = predict(xpred, w);
    plot_all(xs, ys, w, xpred, ypred);
    %plot_cost();
end

function plot_all(xs, ys, w, xpred, ypred)
    xs0 = xs(ys == 0,:);
    xs1 = xs(ys == 1,:);
    syms t;
    sigmoid = @(t)(1 / (1 + exp(-t)));
    
    % plot initial data
    figure;
    hold on;
    plot(xs0, 0, 'bo');
    plot(xs1, 1, 'rx');
    
    % set range for x/y axis
    xmin = min(xs);
    xmax = max(xs);
    xlim([xmin-1 xmax+1]);
    ylim([-0.5 1.5]);
    
    % plot best fit
    delta = 0.1;
    num_pts = (xmax - xmin) / delta + 1;
    wxs = (xmin:delta:xmax)';
    wys = arrayfun(sigmoid, [wxs ones(num_pts, 1)] * w);
    plot(wxs, wys, 'k-');
    
    % plot predicted points
    plot(xpred, ypred, 'go');
end

File: ml/tests/python/scikit_learn/logistic_regression/test1_tumors.py
"""
Mihail Dunaev
May 2017

Simple test that tries to predict if a tumor is malignant (y = 1)
  or benign (y = 0) based on the tumor size (one feature). It uses
  linear logistic regression to do that. Data is in 'tumors.in' in
  input dir.

For logistic regression I used scikit-learn:
  sklearn.linear_model.LogisticRegression

My matlab result
  x0 = 43.2851
  w = 0.6321, -27.3605
Sklearn result
  C = 1 (lambda = 1 in reg)
    w = 0.0474, -1.42906195
    x0 = w(2)/w(1) ~ 30.1489

  C = 1e5 (lambda = 0.00001)
    w = 0.3360, -14.3312
    x0 = 42.6439

  lr.decision_function is not the sigmoid, it's just a linear function
    it should be 'w(1) * x + w(2) = 0', if so it just compares to 0
    and not 0.5 (instead of computing x0).
"""

import matplotlib.pyplot as plt
import numpy as np
from sklearn import linear_model

from common.aux import read_data

def test1_tumors():
    (xs, ys) = read_data("../../../input/tumors.in")

    # train logistic regression model with sklearn
    lr = linear_model.LogisticRegression(C=1e5)
    lr.fit(xs, ys)

    # predict some values
    xpred = np.array([7, 43, 44, 70]).reshape(4, 1);
    ypred = lr.predict(xpred);
    print(ypred.tolist())

    plot_all(xs, ys, lr, xpred, ypred)

def plot_all(xs, ys, lr, xpred, ypred):
    # switch backend - TODO fix this
    plt.switch_backend("TKagg")
    
    N = len(xs)
    xs0 = np.array([xs[i] for i in xrange(N) if ys[i] == 0])
    xs1 = np.array([xs[i] for i in xrange(N) if ys[i] == 1])

    n0 = len(xs0)
    n1 = len(xs1)
    y0 = np.zeros((n0, 1))
    y1 = np.ones((n1, 1))

    # plot initial data
    plt.plot(xs0, y0, "bo")
    plt.plot(xs1, y1, "rx")

    # set range for x/y axis
    xmin = xs.min()
    xmax = xs.max()
    plt.xlim([xmin - 1, xmax + 1])
    plt.ylim([-0.5, 1.5])

    # plot best fit
    num_pts = 100
    wxs = np.linspace(xmin, xmax, num_pts).reshape(num_pts, 1)
    wys = lr.predict_proba(wxs)[:,1]
    plt.plot(wxs, wys, "r-")

    # plot predicted points
    plt.plot(xpred, ypred, color="lime", marker="o", ls="None",
             fillstyle="none")
    plt.show()

test1_tumors()
